programs/mmm/src/instructions/ext_vanilla/sol_ext_fulfill_buy.rs

```diff
@@ -0,0 +1,358 @@
+    #[account(
+        mint::token_program = token_program,
+        constraint = asset_mint.supply == 1 && asset_mint.decimals == 0 @ MMMErrorCode::InvalidTokenMint,
+    )]
+    pub asset_mint: Box<InterfaceAccount<'info, Mint>>,
+    #[account(
+        mut,
+        token::mint = asset_mint,
+        token::authority = payer,
+        token::token_program = token_program,
+    )]
+    pub payer_asset_account: Box<InterfaceAccount<'info, TokenAccount>>,
+    /// CHECK: check in init_if_needed_ata
+    #[account(mut)]
+    pub sellside_escrow_token_account: UncheckedAccount<'info>,
+    /// CHECK: check in init_if_needed_ata
+    #[account(mut)]
+    pub owner_token_account: UncheckedAccount<'info>,
+    /// CHECK: will be used for allowlist checks
+    pub allowlist_aux_account: UncheckedAccount<'info>,
+    #[account(
+        init_if_needed,
+        payer = payer,
+        seeds = [
+            SELL_STATE_PREFIX.as_bytes(),
+            pool.key().as_ref(),
+            asset_mint.key().as_ref(),
+        ],
+        space = SellState::LEN,
+        bump
+    )]
+    pub sell_state: Account<'info, SellState>,
+    pub system_program: Program<'info, System>,
+    pub token_program: Interface<'info, TokenInterface>,
+    pub associated_token_program: Program<'info, AssociatedToken>,
+    pub rent: Sysvar<'info, Rent>,
+    // Remaining accounts
+    // Branch: using shared escrow accounts
+    //   0: m2_program
+    //   1: shared_escrow_account
+    //   2+: transfer hook accounts
+    // Branch: not using shared escrow accounts
+    //   0+: transfer hook accounts
+}
+
+pub fn handler<'info>(
+    ctx: Context<'_, '_, '_, 'info, ExtSolFulfillBuy<'info>>,
+    args: SolFulfillBuyArgs,
+) -> Result<()> {
+    let token_program = &ctx.accounts.token_program;
+    let system_program = &ctx.accounts.system_program;
+    let associated_token_program = &ctx.accounts.associated_token_program;
+    let rent = &ctx.accounts.rent;
+    let pool = &mut ctx.accounts.pool;
+    let sell_state = &mut ctx.accounts.sell_state;
+    let owner = &ctx.accounts.owner;
+    let referral = &ctx.accounts.referral;
+    let payer = &ctx.accounts.payer;
+    let payer_asset_account = &ctx.accounts.payer_asset_account;
+    let asset_mint = &ctx.accounts.asset_mint;
+    let buyside_sol_escrow_account = &ctx.accounts.buyside_sol_escrow_account;
+    let pool_key = pool.key();
+    let buyside_sol_escrow_account_seeds: &[&[&[u8]]] = &[&[
+        BUYSIDE_SOL_ESCROW_ACCOUNT_PREFIX.as_bytes(),
+        pool_key.as_ref(),
+        &[ctx.bumps.buyside_sol_escrow_account],
+    ]];
+    let remaining_accounts = ctx.remaining_accounts;
+
+    check_allowlists_for_mint_ext(
+        &pool.allowlists,
+        &asset_mint.to_account_info(),
+        args.allowlist_aux,
+    )?;
+
+    let (total_price, next_price) =
+        get_sol_total_price_and_next_price(pool, args.asset_amount, true)?;
+    let seller_receives = {
```

Comment:
need to move this as well with the comment, just fyi

Reply:
added